src/h5web/vis-packs/core/utils.ts

```diff
@@ -365,3 +368,39 @@ export function getSliceSelection(
   // Create slice selection string from dim mapping - e.g. [0, 'y', 'x'] => "0,:,:"
   return dimMapping.map((dim) => (isAxis(dim) ? ':' : dim)).join(',');
 }
+
+export function computeSameRatioDomains(
+  canvasRatio: number,
+  imageRatio: number,
+  abscissaDomain: Domain,
+  ordinateDomain: Domain
+): [Domain, Domain] {
+  const x0 = (abscissaDomain[1] - abscissaDomain[0]) / 2;
+  const newAbscissaDomain = abscissaDomain.map(
+    (x) => ((x - x0) * canvasRatio) / imageRatio + x0
+  );
+  if (newAbscissaDomain[1] > abscissaDomain[1]) {
+    return [newAbscissaDomain as Domain, ordinateDomain];
+  }
+
+  const y0 = (ordinateDomain[1] - ordinateDomain[0]) / 2;
+  return [
+    abscissaDomain,
+    ordinateDomain.map(
+      (y) => ((y - y0) * imageRatio) / canvasRatio + y0
+    ) as Domain,
+  ];
+}
```

Comment:
This algorithm is a bit challenging to understand: maybe rename `x0`, `y0`, `newAbscissaDomain`, etc., and add a few comments?